File: backend/src/infrastructure/database/repositories/pg-message.repository.ts
import { injectable, inject } from 'inversify';
import { Pool } from 'pg';
import { IMessageRepository } from '../../../application/repositories/message.repository';
import { Message } from '../../../domain/message';
import { TYPES } from '../../../types';

@injectable()
export class PgMessageRepository implements IMessageRepository {
  constructor(@inject(TYPES.DbPool) private pool: Pool) {}

  async create(messageData: Omit<Message, 'id' | 'createdAt'>): Promise<Message> {
    const { text, authorId, channelId } = messageData;
    const result = await this.pool.query(
      'INSERT INTO messages (text, author_id, channel_id) VALUES ($1, $2, $3) RETURNING *',
      [text, authorId, channelId]
    );
    const row = result.rows[0];
    return {
      id: row.id,
      text: row.text,
      authorId: row.author_id,
      channelId: row.channel_id,
      createdAt: row.created_at,
    };
  }

  async findById(id: string): Promise<Message | null> {
    const result = await this.pool.query('SELECT * FROM messages WHERE id = $1', [id]);
    if (result.rows.length === 0) {
      return null;
    }
    const row = result.rows[0];
    return {
      id: row.id,
      text: row.text,
      authorId: row.author_id,
      channelId: row.channel_id,
      createdAt: row.created_at,
    };
  }

  async findByChannelId(channelId: string): Promise<Message[]> {
    const result = await this.pool.query('SELECT * FROM messages WHERE channel_id = $1 ORDER BY created_at ASC', [channelId]);
    return result.rows.map(row => ({
      id: row.id,
      text: row.text,
      authorId: row.author_id,
      channelId: row.channel_id,
      createdAt: row.created_at,
    }));
  }
}

File: backend/infrastructure/persistence/pg_message_repository.go
package persistence

import (
	"backend/application/repositories"
	"backend/domain"
	"context"
	"database/sql"
	"errors"
)

const (
	createQuery          = `INSERT INTO messages (text, author_id, channel_id) VALUES ($1, $2, $3) RETURNING id, created_at;`
	findByIDQuery        = `SELECT id, text, author_id, channel_id, created_at FROM messages WHERE id = $1;`
	findByChannelIDQuery = `SELECT id, text, author_id, channel_id, created_at FROM messages WHERE channel_id = $1 ORDER BY created_at ASC;`
)

// PgMessageRepository implements the repositories.MessageRepository interface for PostgreSQL.
type PgMessageRepository struct {
	db DBTX
}

// NewPgMessageRepository creates a new PgMessageRepository.
// It accepts a DBTX interface, which can be either a *sql.DB or *sql.Tx.
func NewPgMessageRepository(db DBTX) repositories.MessageRepository {
	return &PgMessageRepository{db: db}
}

func (r *PgMessageRepository) Create(ctx context.Context, message *domain.Message) (*domain.Message, error) {
	err := r.db.QueryRowContext(ctx, createQuery, message.Text, message.AuthorID, message.ChannelID).Scan(&message.ID, &message.CreatedAt)
	if err != nil {
		return nil, err
	}
	return message, nil
}

func (r *PgMessageRepository) FindByID(ctx context.Context, id int) (*domain.Message, error) {
	var message domain.Message
	err := r.db.QueryRowContext(ctx, findByIDQuery, id).Scan(&message.ID, &message.Text, &message.AuthorID, &message.ChannelID, &message.CreatedAt)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &message, nil
}

func (r *PgMessageRepository) FindByChannelID(ctx context.Context, channelID int) ([]*domain.Message, error) {
	rows, err := r.db.QueryContext(ctx, findByChannelIDQuery, channelID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var messages []*domain.Message
	for rows.Next() {
		var message domain.Message
		err := rows.Scan(&message.ID, &message.Text, &message.AuthorID, &message.ChannelID, &message.CreatedAt)
		if err != nil {
			return nil, err
		}
		messages = append(messages, &message)
	}

	if err := rows.Err(); err != nil {
		return nil, err
	}

	return messages, nil
}